ECLIPs/ECIP-draft_succinct-proofs.md

```diff
@@ -0,0 +1,142 @@
+
+FlyClient is that something, and adding an `MMRroot` to each blockheader is the protocol change that will allow FlyClient to work.
+
+On a high level, it allows the verifier to sample a subset of blocks in a tree, rather than a full amount thought a linked list, improving the time-complexit from O(n) to O(log(n)). The sample subset can be made large enough such it is cryptographically infeasible to ever produce a false proof.
+
+With this EIP, PoW can be proven in < 3 megabytes. This will hit a sweet spot, where for instance, mobile wallets, can request the latest cummulative work every time they are opened (no need to even "sync"). Users can finally verify data in a practical way (without a full-node). The proofs are even small enough for IoT devices.
+
+This may also open doors for side-chains, state-channels, plasma, trustless bridges, atomic swaps, and the peer-to-peer light-client-protocols being worked on by the Geth and Parity teams.
+
+This also solves EIPs 210, 1094, 641 - it gives the EVM a way to access to historic blockhashes. A transaction sender simply provides a merkle proof relating the historic hash against any of the last 256 blocks, and the block against its blockhash (available as op `0x40`), with a size of ~864 bytes.
+
+
+## Specification
+<!--The technical specification should describe the syntax and semantics of any new feature. The specification should be detailed enough to allow competing, interoperable implementations for any of the current Ethereum platforms (go-ethereum, parity, cpp-ethereum, ethereumj, ethereumjs, and [others](https://github.com/ethereum/wiki/wiki/Clients)).-->
+
+There is a consensus-breaking rule change proposed here, and then a bit of software around the change will be needed in order to take advantage of the new features. I think it would be best to keep _this_ EIP as just detailing the rule-change, I will summeraize the other tools and link them
+
+### Consensus Rule Change
+
+If `block.number >= ISTANBUL_FORK_NUM` then when constructing the block, set a 16th value `blockHashesRoot` in the header to the `MMRRoot` from the previous block.
+
+This Root is defined by the combining all previous hashes into a [Merkle Mountain Range](https://github.com/juinc/tilap/issues/244) (Peter Todd) using it's block number as leaf index.
+
+Implimentaion of this tree is up to the clients but its fairly simple and requires only an additional ~500mb (on disk) to a full/archive node.
+
+The MMR structure uses a proccess of "bagging the peaks" in order to create a single _root_ hash. However, this may be sub-optimal for our use case. I would like to ammend this to instead define each root as _the hash of the concatonation of the mountain peaks_. This will make building and verifying proofs much simpler without significantly changing the size of the proofs.
```

Comment:
```suggestion
The MMR structure uses a process of "bagging the peaks" in order to create a single _root_ hash. However, this may be sub-optimal for our use case. I would like to amend this to instead define each root as _the hash of the concatenation of the mountain peaks_. This will make building and verifying proofs much simpler without significantly changing the size of the proofs.
```

It's not clear why "it may be sub-optimal for our use case". Is it about performance? Feasibility? Ease of use? Ease of implementation? Size?
In what aspect "hash of concatenation of the mountain peaks" is better than full tree structure (Todd's "bagged peaks" approach)? 
Is the "hash of concatenation" similar to the bagging method described in [grin docs](https://github.com/mimblewimble/grin/blob/master/doc/mmr.md)?

Reply:
this is a good question which i did not describe fully. @tzdybal, from [my js implementation](https://github.com/zmitton/merkle-mountain-range)

> The most elegant way of dealing with merkle-proofs is to basically think of them as a sparse tree in which only some of its nodes are present, and where a proof can be considered invalid iff, while traversing said tree, an undefined node is touched, or a node exists who's children do not hash to itself (more on this below). The second tree structure made from the peaks would not have the same property of being append only and theirfore, would not contain nodes that can be referenced by a perminent node index from the array. This means shared proofs would have to be less effiecient or more complex in some way.

so to what does a proof look like? it is the same data structure as the tree, except the "node database" only contains _some_ of the nodes. When i attempt to `get()` leaves from the tree, I will traverse the necessary nodes, which validate the proof as I go. 

For this to work properly, since i dont have all the nodes, I need to explicitly associate each node with its index position. The problem with "bagging the peaks" is that these "bag nodes" are different. They are not persistent, and do not fit the above abstraction. So in order to support such proofs I would have to pass more information (the  "bag nodes") with the proof.

Not only is this more complex, but when proving _multiple_ leaves at once, the proofs would be larger, because the bag node data can not be shared between leaves (unlike the regular database nodes). I will attempt to createa a picture below

Reply:
A tree with 12 leaves would look like this:
![image](https://user-images.githubusercontent.com/6954889/57118823-d53fd200-6d33-11e9-87fc-2fcbc1482e18.png)
If I wanted to verify the 5th leaf against a known root, I could use only the circled nodes. Also notice that at the moment leaf 5 was inserted, the peaks (needed to verify _its_ merkle root) were the blue circles (6 and 9), which nicely _happen_ to already be part of the later proof. if I also wanted to verify leaf 8, I would just add a few more nodes to the set: 15, 16, 17, and 20.

Using Todd's method, I would need more nodes (for instance a node that is the hash of 14 and 21) - "bag nodes". Those nodes are a separate set. This would complicate the relationship between leaf position, and the proof hashes.

Also, it would make multi-proofs less efficient. Since the peaks are constantly changing, the bag nodes would change too, and could not be completely shared between proofs the same way that the append-only ones are.